File: healthchecks/payment-service/src/main/java/com/varunu28/paymentservice/advice/PaymentControllerAdvice.java
package com.varunu28.paymentservice.advice;

import com.varunu28.paymentservice.exception.DuplicateIdempotencyException;
import com.varunu28.paymentservice.exception.PaymentNotFoundException;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

@ControllerAdvice
public class PaymentControllerAdvice {

    @ExceptionHandler(PaymentNotFoundException.class)
    public ResponseEntity<String> handlePaymentNotFoundException(PaymentNotFoundException ex) {
        return ResponseEntity.status(404).body(ex.getMessage());
    }

    @ExceptionHandler(DuplicateIdempotencyException.class)
    public ResponseEntity<String> handleDuplicateIdempotencyException(DuplicateIdempotencyException ex) {
        return ResponseEntity.status(409).body(ex.getMessage());
    }
}

File: saga/payment-service/src/main/kotlin/com/saga/paymentservice/PaymentControllerAdvice.kt
package com.saga.paymentservice

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler

@ControllerAdvice
class PaymentControllerAdvice {

    @ExceptionHandler(PaymentAmountExceededException::class)
    fun handlePaymentAmountExceedException(e: PaymentAmountExceededException): ResponseEntity<String> {
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.message)
    }
}